File: frontend/src/app/core/services/facture.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import {
  Facture,
  FactureRequest,
  FactureUpdate,
  FactureResponse,
  FactureStatut
} from '../models/facture.model';

@Injectable({
  providedIn: 'root'
})
export class FactureService {
  private apiUrl = `${environment.apiUrl}/api/factures`;

  constructor(private http: HttpClient) {}

  getFactures(): Observable<FactureResponse[]> {
    return this.http.get<FactureResponse[]>(this.apiUrl);
  }

  getFacture(id: number): Observable<FactureResponse> {
    return this.http.get<FactureResponse>(`${this.apiUrl}/${id}`);
  }

  getFacturesByClient(clientId: number): Observable<FactureResponse[]> {
    return this.http.get<FactureResponse[]>(`${this.apiUrl}/client/${clientId}`);
  }

  getOverdueFactures(): Observable<FactureResponse[]> {
    return this.http.get<FactureResponse[]>(`${this.apiUrl}/overdue`);
  }

  getFacturesByStatus(status: FactureStatut): Observable<FactureResponse[]> {
    return this.http.get<FactureResponse[]>(`${this.apiUrl}/status/${status}`);
  }

  createFacture(facture: FactureRequest): Observable<FactureResponse> {
    return this.http.post<FactureResponse>(this.apiUrl, facture);
  }

  updateFacture(id: number, facture: FactureUpdate): Observable<FactureResponse> {
    return this.http.put<FactureResponse>(`${this.apiUrl}/${id}`, facture);
  }

  markAsPaid(id: number): Observable<void> {
    return this.http.patch<void>(`${this.apiUrl}/${id}/paid`, {});
  }

  deleteFacture(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}

File: backend/factureservice/src/main/java/com/example/factureservice/service/FactureService.java
package com.example.factureservice.service;

import com.example.factureservice.dto.FactureRequest;
import com.example.factureservice.dto.FactureResponse;
import com.example.factureservice.dto.FactureUpdate;

import java.util.List;
import java.util.Optional;

public interface FactureService {
    List<FactureResponse> getAllFactures();
    Optional<FactureResponse> getFactureById(Long id);
    Optional<FactureResponse> getFactureByNumero(String numero);
    List<FactureResponse> getFacturesByClientId(Long clientId);
    List<FactureResponse> getFacturesByStatut(String statut);
    FactureResponse createFacture(FactureRequest request);
    Optional<FactureResponse> updateFacture(Long id, FactureUpdate update);
    void deleteFacture(Long id);
    void markAsPaid(Long id);
    List<FactureResponse> getOverdueFactures();
}

